packages/libraries/router/src/persisted_documents.rs

```diff
@@ -0,0 +1,790 @@
+            .timeout(Duration::from_secs(config.request_timeout.unwrap_or(15)))
+            .build()
+            .expect("Failed to create reqwest client");
+
+        let cache_size = config.cache_size.unwrap_or(1000);
+        let cache = Arc::new(Mutex::new(LruCache::<String, String>::new(
+            NonZeroUsize::new(cache_size).unwrap(),
+        )));
+
+        Self {
+            agent,
+            cache,
+            config: config.clone(),
+        }
+    }
+
+    /// Extracts the document id from the request body.
+    /// In case of a parsing error, it returns the error.
+    /// This will also try to parse other GraphQL-related (see `original_req`) fields in order to 
+    /// pass it to the next layer.
+    fn extract_document_id(
+        body: &hyper::body::Bytes,
+    ) -> Result<ExpectedBodyStructure, PersistedDocumentsError> {
+        serde_json::from_slice::<ExpectedBodyStructure>(&body)
+            .map_err(PersistedDocumentsError::FailedToParseBody)
+    }
+
+    /// Resolves the document from the cache, or from the CDN
+    async fn resolve_document(&self, document_id: &str) -> Result<String, PersistedDocumentsError> {
+        let cached_record = self.cache.lock().await.get(document_id).cloned();
+
+        match cached_record {
+            Some(document) => {
+                debug!("Document {} found in cache: {}", document_id, document);
+
+                return Ok(document);
+            }
+            None => {
+                debug!(
+                    "Document {} not found in cache. Fetching from CDN",
+                    document_id
+                );
+                let cdn_document_id = str::replace(document_id, "~", "/");
+                let cdn_artifact_url = format!("{}/apps/{}", self.config.endpoint.as_ref().unwrap(), cdn_document_id);
+                info!(
+                    "Fetching document {} from CDN: {}",
+                    document_id, cdn_artifact_url
+                );
+                let cdn_response = self
+                    .agent
+                    .get(cdn_artifact_url)
+                    .header("X-Hive-CDN-Key", self.config.key.as_ref().unwrap())
+                    .send()
+                    .await;
+
+                match cdn_response {
+                    Ok(response) => {
+                        if response.status().is_success() {
+                            let document = response.text().await.unwrap();
+                            debug!(
+                                "Document fetched from CDN: {}, storing in local cache",
+                                document
+                            );
+                            self.cache
+                                .lock()
+                                .await
+                                .put(document_id.into(), document.clone());
+
+                            return Ok(document);
+                        }
```

Comment:
this will likely reduce the throughput of the incoming requests to the gateway, bc we're locking the cache on every successful request, which means that if the gateway can process 10 requests at once (concurrently), this will be a bottleneck because each request will wait for the other to release the `cache` lock for it to be acquired by the other thread.

maybe `RwLock` or `dashmap` can be used for the cache instead of a `Mutex`?

Reply:
Hmm, good point, but this one is tricky, since we'll have to get `rwLock.write()` also for `cache.get` calls, because the `LruCache` needs `&mut self` in order to run `.get` calls (so write access is required, to update the LRU cache).

Reply:
See https://stackoverflow.com/questions/69617603/using-rwlock-read-on-a-mutable-struct

Reply:
hmmm...you're right it's not as straightforward, let me see what I can do, I'll experiment with it